go_lib/cloud-data/reconciler.go

```diff
@@ -0,0 +1,271 @@
+// Copyright 2023 Flant JSC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cloud_data
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	log "github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/dynamic"
+
+	"github.com/deckhouse/deckhouse/go_lib/cloud-data/apis/v1alpha1"
+)
+
+type Discoverer interface {
+	InstanceTypes(ctx context.Context) ([]v1alpha1.InstanceType, error)
+}
+
+type Reconciler struct {
+	cloudRequestErrorMetric   *prometheus.GaugeVec
+	updateResourceErrorMetric *prometheus.GaugeVec
+
+	discoverer    Discoverer
+	checkInterval time.Duration
+	listenAddress string
+	logger        *log.Entry
+	k8sClient     dynamic.Interface
+}
+
+func NewReconciler(
+	discoverer Discoverer,
+	listenAddress string,
+	interval time.Duration,
+	logger *log.Entry,
+	k8sClient dynamic.Interface,
+) *Reconciler {
+	return &Reconciler{
+		checkInterval: interval,
+		listenAddress: listenAddress,
+		discoverer:    discoverer,
+		logger:        logger,
+		k8sClient:     k8sClient,
+	}
+}
+
+func (c *Reconciler) Start() {
+	defer c.logger.Infoln("Stop cloud data discoverer fully")
+
+	c.logger.Infoln("Start cloud data discoverer")
+	c.logger.Infoln("Address: ", c.listenAddress)
+	c.logger.Infoln("Checks interval: ", c.checkInterval)
+
+	// channels to stop converge loop
+	shutdownAllCh := make(chan struct{})
```

Comment:
NIT: I guess we can use `rootCtx` with cancel instead of this channel

Reply:
done